### PR TITLE
Update PUDL to pandas 2.0

### PR DESCRIPTION
**Pull request by zaneselvans**
# PR Overview

# PR Checklist
- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [ ] Make sure you've included good docstrings.
- [ ] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [ ] Include unit tests for new functions and classes.
- [ ] Defensive data quality/sanity checks in analyses & data processing functions.
- [ ] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [ ] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

### CODE REVIEW

**Comment by zaneselvans** [comment]
After bumping SQLAlchemy to v2 I got some additional unit test failures in the `pudl_sqlite_io_manager` tests dealing with views and the foreign key checks.

**Comment by zaneselvans** [comment]
@nelsonauner I fixed another couple of small issues on this branch, and ended up reporting what seems to be a pandas bug: https://github.com/pandas-dev/pandas/issues/54399 

Now the integration tests are failing on a divide-by-zero error in the FERC Form 1 calculation reconciliations.

File: test/unit/io_managers_test.py [hunk 1]
```diff
@@ -218,6 +218,7 @@ def test_error_when_handling_view_without_metadata(pudl_sqlite_io_manager_fixtur
         pudl_sqlite_io_manager_fixture.handle_output(output_context, sql_stmt)
 
 
+@pytest.mark.xfail(reason="SQLAlchemy is not finding the view. Debug or remove.")
```
Comment (zaneselvans) [inline test/unit/io_managers_test.py:221]: I'm inclined to leave this `xfail()` for now, and see if it's easy to fix when integrating SQLAlchemy 2.0 (which we should do as soon as pandas 2.0 is merged in)

**Review by zaneselvans — COMMENTED** [review]
I've made comments throughout the PR about why things were changed (if there's no comment it's probably a vanilla data type issue).

There are several instances of `df.convert_dtypes()` that Nelson introduced, which I think should *probably* be `apply_pudl_dtypes()` which is more selective / specific.

There's one unit test with mixed date formats that I still need to update. Mixed dates aren't automatically parsed by Pandas any more, and it turns out this didn't come up anywhere else in PUDL, so I just need to make all the date formats identical and then I can un-xfail that test.

There's one SQLAlchemy error coming up with respect to the SQLite Views, which we aren't using right now.  I've left it xfail for the moment, and thought we might try and fix it when we update to SQLAlchemy 2.0 (which we should do as soon as pandas 2.0 is merged in).

The full ETL and the full data integration tests all pass locally, so it seems like we haven't had any significant impacts on the data contents. 🤞🏼

**Comment by zaneselvans** [comment]
Hey FYI @nelsonauner I went ahead and chased down the remaining pandas 2.0 issues in the full ETL pipeline and integration tests, so I think we're about ready to merge this in!

File: src/pudl/extract/eia_bulk_elec.py [hunk 2]
```diff
@@ -75,11 +75,12 @@ def _parse_data_column(elec_df: pd.DataFrame) -> pd.DataFrame:
             )
         else:
             data_df.loc[:, "date"] = pd.to_datetime(
-                data_df.loc[:, "date"], infer_datetime_format=True, errors="raise"
+                data_df.loc[:, "date"], errors="raise"
             )
         data_df["series_id"] = elec_df.at[idx, "series_id"]
         out.append(data_df)
     out = pd.concat(out, ignore_index=True, axis=0)
+    out = out.convert_dtypes()
```
Comment (zaneselvans) [inline src/pudl/extract/eia_bulk_elec.py:83]: Tried changing this `convert_dtypes()` to `apply_pudl_dtypes()` and it failed, so leaving it like this.

File: test/unit/extract/eia_bulk_elec_test.py [hunk 3]
```diff
@@ -65,7 +65,7 @@ def test__parse_data_column(elec_txt_dataframe):
                 4371683.38189,
             ],
         },
-    )
+    ).convert_dtypes()
```
Comment (zaneselvans) [inline test/unit/extract/eia_bulk_elec_test.py:68]: Tried changing this `convert_dtypes()` to `apply_pudl_dtypes()` and it failed, so leaving it like this.

File: src/pudl/transform/ferc1.py [hunk 4]
```diff
@@ -3926,20 +3927,16 @@ def prep_header_fuel_and_plant_types(
...
-            [
-                "utility_id_ferc1",
-                "report_year",
-                (df["row_type"] == "header").cumsum(),
-            ]
-        )
-        # Forward fill based on headers
-        df.loc[df["row_type"] != "note", "header"] = header_groups.header.ffill()
+        df = df.reset_index(drop=True)
+        df["header_group"] = (df["row_type"] == "header").cumsum()
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/pudl/transform/ferc1.py:3931]: Do you know why the `groupby` wasn't having a problem in pandas 1.0?

File: src/pudl/extract/eia_bulk_elec.py (hunk 2 above)
Comment (zschira) [reply src/pudl/extract/eia_bulk_elec.py:83]: I don't think it's a big deal, but would it be worth investigating why `apply_pudl_dtypes()` is failing and fixing it, so we can consistently use `apply_pudl_dtypes`?

**Review by zschira — APPROVED** [review]
This looks good, thanks for doing this! I don't have anything blocking, just a few questions.

File: test/unit/io_managers_test.py (hunk 1 above)
Comment (zaneselvans) [reply test/unit/io_managers_test.py:221]: I agree. We're not using any SQL views right now, but if we do want to use them at some point we need to make them work with SQLAlchemy 2.0, so I think in that PR we need to either figure out how to make this work or explicitly decide we're not going to use views (which is what we've de facto chosen to do thus far, given the complexities around defining their metadata correctly)

**Pull request merged**